### PR TITLE
fix several podman events issues

### LINKED ISSUE #15688: podman events missbehaving when there are zero events
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

podman events --stream=false is used for many test to make sure that we do not hang in the tests.
However as I found out while working on https://github.com/containers/podman/pull/15673 that --events-backend=file hangs if the event log file does not exists.
For --events-backend=none it also hangs for some reason? I would expect an error since there are no events.
Lastly with --events-backend=journald it just errors out right away when there are no events.

**Steps to reproduce the issue:**
file, as rootless:
```
$ rm $XDG_RUNTIME_DIR/libpod/tmp/events/events.log
$ podman events --events-backend=file --stream=false <--just hangs until you create events or create the file
```

journald:
first empty the journal
```
$ sudo journalctl --rotate
$ sudo journalctl --vacuum-time=1s
```
```
$ podman events --events-backend=journald --stream=false
Error: failed to get journal cursor: failed to get cursor: cannot assign requested address
$ podman events --events-backend=journald <--same error without stream=false
Error: failed to get journal cursor: failed to get cursor: cannot assign requested address
```

none:
```
$ podman events --events-backend=none --stream=false  <-- hangs without error
$ podman events --events-backend=none  <-- also hangs without error
```

**Describe the results you expected:**
The file backend with --stream=false should not hang as this can/will break tests.

The journald backend just not fail when there are no events.

The podman events with none backend should just error, there is no reason to ever run this command.

**Output of `podman version`:**

```
latest main
```

### PR DESCRIPTION
**Pull request by Luap99**
### fix hang with podman events file logger
podman --events-backend file events --stream=false should never hang. The
problem is that our tail library will wait for the file to be created
which makes sense when we do not run with --stream=false. To fix this we
can just always create the file when the logger is initialized. This
would also help to report errors early on in case the file is not
accessible.

### event backend none: return an error when reading events
podman --events-backend none events should return with an error since it
will never be able to actually list events.

### event backend journald: fix problem with empty journal

Currently podman events will just fail with `Error: failed to get journal
cursor: failed to get cursor: cannot assign requested address` when the
journal contains zero podman events.

The problem is that we are using the journal accessors wrong. There is no
need to call GetCursor() and compare them manually. The Next() return an
integer which tells if it moved to the next or not. This means the we can
remove GetCursor() which would fail when there is no entry.

This also includes another bug fix. Previously the logic called Next()
twice for the first entry which caused us to miss the first entry.

To reproduce this issue you can run the following commands:
```
sudo journalctl --rotate
sudo journalctl --vacuum-time=1s
```
Note that this will delete the full journal.

Now run podman events and it fails but with this patch it works.
Now generate a single event, i.e. podman pull alpine, and run
podman events --until 1s.

I am not sure how to get a reliable test into CI, I really do not want
to delete the journal and developer or CI systems.

### libpod: runtime newEventer() cleanup

There is no reason to create a new eventer every time. The libpod runtime
already has one attached which should be used instead.

Fixes #15688

#### Does this PR introduce a user-facing change?

```release-note
podman --events-backend file events --stream=false will now no longer hang when the event file does not exists.
podman --events-backend none events now returns an error because it can never list events.
podman --events-backend journald will now work with an empty journal and no longer skip the first event.
```

### CODE REVIEW

File: cmd/podman/system/events.go [hunk 1]
```diff
@@ -99,25 +99,28 @@ func eventsCmd(cmd *cobra.Command, _ []string) error {
...
 
-	for event := range eventChannel {
-		switch {
-		case event == nil:
-			// no-op
-		case doJSON:
-			jsonStr, err := event.ToJSONString()
-			if err != nil {
-				return err
+	for {
```
Comment (vrothberg) [inline cmd/podman/system/events.go:102]: The changes here look unrelated to the commit.

File: libpod/events/journal_linux.go [hunk 2]
```diff
@@ -112,57 +112,16 @@ func (e EventJournalD) Read(ctx context.Context, options ReadOptions) error {
...
-			return nil
-		default:
-			// fallthrough
-		}
-
-		if _, err := j.Next(); err != nil {
-			return fmt.Errorf("failed to move journal cursor to next entry: %w", err)
-		}
-		newCursor, err := j.GetCursor()
+		entry, err := getNextEntry(ctx, j, options.Stream, untilTime)
```
Comment (vrothberg) [inline libpod/events/journal_linux.go:116]: Nit: bug fix + refactor in the same commit is hard to review.

**Review by vrothberg — COMMENTED** [review]
Nice work, @Luap99!

Changes LGTM

File: cmd/podman/system/events.go (hunk 1 above)
Comment (Luap99) [reply cmd/podman/system/events.go:102]: They are not, for loop or event channel is blocking. Therefore we can never read the error channel.

File: libpod/events/logfile.go [hunk 3]
```diff
@@ -27,6 +27,18 @@ type EventLogFile struct {
...
+// newLogFileEventer creates a new EventLogFile eventer
+func newLogFileEventer(options EventerOptions) (*EventLogFile, error) {
+	// We have to make sure the file is created otherwise reading events will hang.
+	// https://github.com/containers/podman/issues/15688
+	fd, err := os.OpenFile(options.LogFilePath, os.O_RDONLY|os.O_CREATE, 0700)
+	if err != nil {
+		return nil, err
+	}
+	_ = fd.Close()
+	return &EventLogFile{options: options}, nil
```
[the remark this answers is not in the document]
Comment (Luap99) [reply libpod/events/logfile.go:39]: I can add this but honestly do we want to fail basically all podman commands for this?

**Comment by edsantiago** [comment]
Since you need to resubmit anyway... could I beg you to include something like this?
```diff
diff --git a/docs/source/markdown/podman.1.md b/docs/source/markdown/podman.1.md
index 3b3974dcc..e992377cf 100644
--- a/docs/source/markdown/podman.1.md
+++ b/docs/source/markdown/podman.1.md
@@ -43,8 +43,8 @@ Remote connections use local containers.conf for default.                                                                                    
 #### **--events-backend**=*type*
 
 Backend to use for storing events. Allowed values are **file**, **journald**, and
-**none**. When *file* is specified, the events are stored under a subdirectory
-of the *tmpdir* location (see **--tmpdir** below).
+**none**. When *file* is specified, the events are stored
+in \$XDG\_RUNTIME\_DIR/libpod/tmp/events/events.log.
 
 #### **--help**, **-h**
 ```
This would address my comment about debug logging. Basically, I wasted a lot of time yesterday looking for a (nonexistent) log or lock file "somewhere under a tmpdir".

**Comment by Luap99** [comment]
> Since you need to resubmit anyway... could I beg you to include something like this?
> 
> ```diff
> diff --git a/docs/source/markdown/podman.1.md b/docs/source/markdown/podman.1.md
> index 3b3974dcc..e992377cf 100644
> --- a/docs/source/markdown/podman.1.md
> +++ b/docs/source/markdown/podman.1.md
> @@ -43,8 +43,8 @@ Remote connections use local containers.conf for default.                                                                                    
>  #### **--events-backend**=*type*
>  
>  Backend to use for storing events. Allowed values are **file**, **journald**, and
> -**none**. When *file* is specified, the events are stored under a subdirectory
> -of the *tmpdir* location (see **--tmpdir** below).
> +**none**. When *file* is specified, the events are stored
> +in \$XDG\_RUNTIME\_DIR/libpod/tmp/events/events.log.
>  
>  #### **--help**, **-h**
> ```
> 
> This would address my comment about debug logging. Basically, I wasted a lot of time yesterday looking for a (nonexistent) log or lock file "somewhere under a tmpdir".


sure

As it looks the test found another bug. They use the default log file location even when --root and --runroot and --tmpdir are set.

**Comment by edsantiago** [comment]
Because the diff isn't obvious: test is expecting untag+remove as last two events, is getting untag+untag instead.

**Review by vrothberg — COMMENTED** [review]
LGTM

**Comment by Luap99** [comment]
@containers/podman-maintainers PTAL

File: docs/source/markdown/podman.1.md [hunk 4]
```diff
@@ -158,7 +158,7 @@ On remote clients, including Mac and Windows (excluding WSL2) machines, logging
 
 #### **--tmpdir**
 
-Path to the tmp directory, for libpod runtime content.
+Path to the tmp directory, for libpod runtime content. Defaults to `$XDG\_RUNTIME\_DIR/libpod/tmp` as rootless and `run/libpod/tmp` as rootful.
```
Comment (edsantiago) [inline docs/source/markdown/podman.1.md:161]: Should be `/run` (missing slash)?

**Review by edsantiago — CHANGES_REQUESTED** [review]
Trivial nit in man page, but am blocking because, could you check your commit list please? I think you accidentally adopted 72e715a11

**Comment by Luap99** [comment]
> Trivial nit in man page, but am blocking because, could you check your commit list please? I think you accidentally adopted [72e715a](https://github.com/containers/podman/commit/72e715a1109426114ef054042be28014380a246d)

No that commit is required to vendor in new c/common (breaking changes)

**Review by edsantiago — APPROVED** [review]

**Comment by edsantiago** [comment]
>No that commit is required to vendor in new c/common (breaking changes)

Got it. For benefit of future spelunkers, that was originally #15723, which was closed in favor of getting absorbed here.
